### PR TITLE
ranking: add IDF to BM25 score calculation

### PR DESCRIPTION
**Pull request by stefanhengl**
So far, we didn't include IDF in our BM25 score function. Zoekt uses a
trigram index and hence doesn't compute document frequency during
indexing. We could add this information to the index, but it is not
immediately obvious how to tokenize code in a way that is compatible
with tokens from a natural language query.

Here we calulate the document frequency at query time under the
assumption that we visit all documents containing any of the query terms.

Notes:
Also fixed an off-by-1 bug with how we count documents.

Test plan:
- Updated unit test
- Context evaluation results are slightly worse with a decrease from 64/89 to 63/89

### CODE REVIEW

File: score.go [hunk 1]
```diff
@@ -141,8 +138,7 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
 	// Compute the file length ratio. Usually the calculation would be based on terms, but using
 	// bytes should work fine, as we're just computing a ratio.
 	fileLength := float64(d.boundaries[doc+1] - d.boundaries[doc])
-	numFiles := len(d.boundaries)
```
Comment (stefanhengl) [inline score.go:144]: `d.boundaries` has `numFiles+1` entries, the last item being the end of the last file.

**Comment by stefanhengl** [comment]
**golden queries evals**

before
```
Breakdown by class:
Find symbol     9/10
Find string     2/2
Explain file    2/2
Explain concept 4/5
Check dependency        1/2
Find logic      31/43
Gather information      12/16
Changelog       0/2
Ownership       2/2
How-to  1/1
Foreign language        0/2
Long request    0/2

Breakdown by file type:
TSX     4/5
Golang  17/24
Typescript      27/38
C++     1/3
Markdown        7/10
Graphql 1/1
JSON    1/1
Go      3/4
Codeowners      2/2
Python  1/1

Combined recall 64/89
```

after
```
Breakdown by class:
Find symbol     9/10
Find string     2/2
Explain file    2/2
Explain concept 4/5
Check dependency        1/2
Find logic      31/43
Gather information      11/16
Changelog       0/2
Ownership       2/2
How-to  1/1
Foreign language        0/2
Long request    0/2

Breakdown by file type:
TSX     4/5
Golang  17/24
Typescript      26/38
C++     1/3
Markdown        7/10
Graphql 1/1
JSON    1/1
Go      3/4
Codeowners      2/2
Python  1/1

Combined recall 63/89
```

File: score.go [hunk 2]
```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
...
+	fileMatches []FileMatch
+	tfs         []termFrequencyScore
+}
+
+func (m *fileMatchesWithScores) addFileMatch(fm FileMatch, tfs termFrequencyScore) {
+	m.fileMatches = append(m.fileMatches, fm)
+	m.tfs = append(m.tfs, tfs)
+}
+
+func (m *fileMatchesWithScores) scoreFilesUsingBM25(df termDocumentFrequency, documentCount int) {
```
Comment (jtibshirani) [inline score.go]: As a small tweak, we could collect the term frequencies instead of scores, then perform the full BM25 calculation in `scoreFilesUsingBM25`. I like that because it keeps the BM25 calculation in one place, which makes it easier to document and understand the formula. Also we could only have one data type that's `map[string]int` (maybe renaming `termDocumentFrequency` -> `frequencies`) ?

Also we could move the `fileMatch.Debug` calculation to this method to report both IDF and the term frequency component?

File: score.go [hunk 3]
```diff
@@ -115,16 +108,20 @@ func (d *indexData) scoreFile(fileMatch *FileMatch, doc uint32, mt matchTree, kn
...
+// calculateTermFrequencyScore computes the TF score per term for the file match
+// according to BM25, the most common scoring algorithm for text search:
+// https://en.wikipedia.org/wiki/Okapi_BM25. We defer the calculation of the
+// full bm25 score to after we have finished searching the shard, because we can
+// only calculate the inverse document frequency (idf) after we have seen all
+// documents.
 //
-// Filename matches count twice as much as content matches. This mimics a common text search strategy where you
-// 'boost' matches on document titles.
+// Filename matches count more than content matches. This mimics a common text
```
Comment (jtibshirani) [inline score.go:113]: Thanks for catching this.

File: score.go [hunk 4]
```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
...
+
+	tfs := make(termFrequencyScore)
+
+	for term, freq := range termFreqs {
 		tf := float64(freq)
 		sumTf += tf
-		score += ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
+
+		// Invariant: the keys of df are the union of the keys of tfs over all files.
+		df[term] += 1
```
Comment (jtibshirani) [inline score.go:129]: Double-checking: we don't have any setting that allows within-shard parallelism right? So there could never be contention on this `df` map.

File: score.go [hunk 5]
```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
...
+
+// termDocumentFrequency is a map "term" -> "number of documents that contain the term"
+type termDocumentFrequency map[string]int
+
+// termFrequencyScore is a map "term" -> "term frequency score"
+type termFrequencyScore map[string]float64
+
+// fileMatchesWithScores is a helper type that is used to store the file matches
+// along with internal scoring information.
+type fileMatchesWithScores struct {
```
Comment (jtibshirani) [inline score.go]: I agree with you that this is a little awkward! If we can't put this on `FileMatch` Could we add the term frequencies to `SearchResult` instead?

**Comment by jtibshirani** [comment]
This looks good! Even if it doesn't improve the results over our current baseline, I feel good about this since we should really be implementing BM25 properly. Also we've definitely overfit to the golden queries by now, so I'm not too worried if we don't see a clear improvement.

I'd love to see these evals in particular:
* Compare to latest golden queries snapshot
* Compare to golden queries *without* my latest change to identify and search symbol definitions
* Run on CodeSearchNet too

File: score.go [hunk 6]
```diff
@@ -153,12 +149,56 @@ func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands [
...
-	for _, freq := range termFreqs {
+
+	tfs := make(termFrequencyScore)
+
+	for term, freq := range termFreqs {
 		tf := float64(freq)
 		sumTf += tf
-		score += ((k + 1.0) * tf) / (k*(1.0-b+b*L) + tf)
+
+		// Invariant: the keys of df are the union of the keys of tfs over all files.
```
Comment (jtibshirani) [inline score.go]: As we discussed, this is tricky! I think for now we should loudly document this on the `UseBM25Scoring` option, so users know how it works.

In a follow-up, I'd love to introduce a new Zoekt query type like "TextQuery" that takes a list of terms, creates a disjunction, and applies BM25. Then users could *only* use BM25 with this query, and not accidentally use it with other types.

**Comment by stefanhengl** [comment]
> This looks good! Even if it doesn't improve the results over our current baseline, I feel good about this since we should really be implementing BM25 properly. Also we've definitely overfit to the golden queries by now, so I'm not too worried if we don't see a clear improvement.
> 
> I'd love to see these evals in particular:
> 
> * Compare to latest golden queries snapshot

see above

> * Compare to golden queries _without_ my latest change to identify and search symbol definitions

**golden queries evals**

This PR with `Sourcegraph@4f465c5`, which is the parent commit of your latest change to identify and search symbol definitions, see [here](https://github.com/sourcegraph/sourcegraph/commit/91518a1ef7f5b9a6627e5b7c52c9ea5c068551fa)
```
Breakdown by class:
Find symbol     9/10
Find string     2/2
Explain file    2/2
Explain concept 4/5
Check dependency        1/2
Find logic      31/43
Gather information      11/16
Changelog       0/2
Ownership       2/2
How-to  1/1
Foreign language        0/2
Long request    0/2

Breakdown by file type:
TSX     4/5
Golang  17/24
Typescript      26/38
C++     1/3
Markdown        7/10
Graphql 1/1
JSON    1/1
Go      3/4
Codeowners      2/2
Python  1/1

Combined recall 63/89
```

File: score.go (hunk 5 above)
Comment (stefanhengl) [reply score.go]: I pushed a new version in which I add the term frequencies to `FileMatch`. I had to tweak the tests quite a bit to make `quick.Check` happy. Let me know what you think.

**Comment by stefanhengl** [comment]
**CodeSearchNet**

`Sourcegraph@4f465c5`

Before
```
Recall (files)  91/99
Recall (chunks) 75/99
Average chunk overlap   0.89
```

After
```
Recall (files)  89/99
Recall (chunks) 77/99
Average chunk overlap   0.88
```

File: score.go (hunk 4 above)
Comment (stefanhengl) [reply score.go:129]: Correct, we don't have within-shard parallelism.

File: api_proto.go [hunk 7]
```diff
@@ -83,6 +85,28 @@ func (m *FileMatch) ToProto() *proto.FileMatch {
 	}
 }
 
+func (*FileMatch) Generate(rng *rand.Rand, _ int) reflect.Value {
```
Comment (jtibshirani) [inline api_proto.go]: Do these generate methods usually live on the proto file? Maybe they should be in a test utils file instead?

File: score.go [hunk 8]
```diff
@@ -111,54 +104,82 @@ func (d *indexData) scoreFile(fileMatch *FileMatch, doc uint32, mt matchTree, kn
...
+// search strategy where you 'boost' matches on document titles.
+func calculateTermFrequency(fileMatch *FileMatch, cands []*candidateMatch, df termDocumentFrequency) {
 	// Treat each candidate match as a term and compute the frequencies. For now, ignore case
 	// sensitivity and treat filenames and symbols the same as content.
 	termFreqs := map[string]int{}
 	for _, cand := range cands {
 		term := string(cand.substrLowered)
 
+		if _, ok := termFreqs[term]; !ok {
+			df[term] += 1
```
Comment (jtibshirani) [inline score.go]: I just noticed -- shouldn't we only be increasing `df` once per file, instead of once per candidate match? We consider a file to be a "document" for the BM25 calculation, not a chunk match.

File: score.go (hunk 5 above)
Comment (jtibshirani) [reply score.go]: Definitely a bit tricky! I don't have a better idea. Tagging @keegancsmith in case he has a strong opinion here.

File: api_proto.go (hunk 7 above)
Comment (stefanhengl) [reply api_proto.go]: There is already another generator method there,  that's why I put them there.

File: eval.go [hunk 9]
```diff
@@ -339,16 +348,20 @@ nextFileMatch:
 		repoMatchCount += len(fileMatch.LineMatches)
 		repoMatchCount += matchedChunkRanges
 
-		if opts.DebugScore {
-			fileMatch.Debug = fmt.Sprintf("score:%.2f <- %s", fileMatch.Score, fileMatch.Debug)
-		}
```
Comment (keegancsmith) [inline eval.go:344]: why did you remove this?

File: eval.go [hunk 10]
```diff
@@ -318,7 +322,12 @@ nextFileMatch:
 		}
 
 		if opts.UseBM25Scoring {
-			d.scoreFileUsingBM25(&fileMatch, nextDoc, finalCands, opts)
+			// For BM25 scoring, the calculation of the score is split in two parts. Here we
+			// calculate the term frequencies for the current document and update the
+			// document frequencies. Since we don't store document frequencies in the index,
+			// we have to defer the calculation of the final BM25 score to after the whole
+			// shard has been processed.
+			calculateTermFrequency(&fileMatch, finalCands, df)
```
Comment (keegancsmith) [inline eval.go]: can't we make another slice that has the invariant that `scoreData[i] belongs to res.Files[i]`. Adding private fields to our public API feels very wrong.

File: score.go (hunk 8 above)
Comment (stefanhengl) [reply score.go]: The check `if _, ok := termFreqs[term]; !ok {` should make sure we only increase it once per document. We can pull that out of the loop to make it more obvious at the cost of an additional loop. However, there shouldn't be too many unique terms so it's probably fine. I also added a new unit case.

**Review by keegancsmith — COMMENTED** [review]
added an idea to avoid the public api changes.

File: eval.go (hunk 10 above)
Comment (stefanhengl) [reply eval.go]: you prefer [af9a0a4](https://github.com/sourcegraph/zoekt/pull/788/commits/af9a0a4bc861750531f2ee90ddac53feaedeb7cc)? 

>> Adding private fields to our public API feels very wrong.

We probably should have internal versions of our structs and convert at the API boundary. I don't feel that strongly about adding private fields though if it helps to keep the code simple.

Comment (stefanhengl) [reply eval.go]: I reverted the changes to the API/`FileMatch` and went back to something similar to the first version, but with the BM25 calculation all in one function. I think it is a good trade-off. A ran the evals again without change.

File: eval.go (hunk 9 above)
Comment (stefanhengl) [reply eval.go:344]: I just moved it to the respective scoring function, see `scoreFile` or `scoreFilesUsingBM25`.

File: eval.go [hunk 11]
```diff
@@ -339,16 +351,26 @@ nextFileMatch:
...
 
-		if opts.DebugScore {
-			fileMatch.Debug = fmt.Sprintf("score:%.2f <- %s", fileMatch.Score, fileMatch.Debug)
-		}
-
+		// Invariant: tfs[i] belongs to res.Files[i]
+		tfs = append(tfs, termFrequency{
+			doc: nextDoc,
+			tf:  tf,
+		})
```
Comment (keegancsmith) [inline eval.go]: minor: should we only do this if opts.UseBM25Scoring

**Review by keegancsmith — APPROVED** [review]

**Review by jtibshirani — APPROVED** [review]
Nice! Thanks for all the iterations.
